### ISSUE TITLE
Kubelet eats lots of memory and cpu

### ISSUE THREAD

**Issue by ssergiienko** [kubernetes/kubernetes#15644]
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                  
17083 root      20   0 2050468 936012   7708 S 178.7 49.8  14:16.87 kubelet                                                                                                                  
31383 root      20   0  192248   9072   4636 S   9.6  0.5   0:33.87 flanneld                                                                                                                 
  418 root      20   0   63848  14172  11904 S   3.3  0.8   0:26.42 systemd-journal
```

I reported almost same bug for flannel and found that problem appears when node's system time is in far future from master on which etcd is running. This happens after restore cluster from snapshot. It looks like kubelet has this bug too.

Link to my bug for flanneld (for more description): https://github.com/coreos/flannel/issues/337

Possible this https://github.com/kubernetes/kubernetes/issues/15528 bug may be for the same reason.

**Comment by yujuhong** [comment]
Previous issues about time skew: #6159, #1651

**Comment by yujuhong** [comment]
> Previous issues about time skew in the cluster: #6159, #1651

The conclusion from the old issues was that ntpd should be part of the node requirement. @davidopp @dchen1107 @erictune, does the conclusion still hold today?

**Comment by dchen1107** [comment]
cc/ @yifan-gu @jonboulle

**Comment by jonboulle** [comment]
cross-linking to coreos/flannel#340

**Comment by yujuhong** [comment]
@ssergiienko, what OS distro and version do you use?

**Comment by ssergiienko** [comment]
@yujuhong CentOS Linux release 7.1.1503 (Core)
Kubernetes 1.0.3-0.1.git61c6ac5
Kubernetes are with our own patches but they are related to replication-controller and they are small

**Comment by yujuhong** [comment]
@ssergiienko, thanks for the information.

We are aware that some features rely on sync'd clocks in the cluster to function correctly, and I don't see us attempt to handle time skew in near future. 
The best solution for you might be to run ntpd on your nodes to ensure the clocks are in sync.

That said, perhaps kubelet can warn or fail more gracefully. @ssergiienko, are you able to reproduce this issue reliably? I tried modifying the date on some nodes in my cluster yesterday, but weren't able to cause any abnormal cpu usage.

**Comment by ssergiienko** [comment]
We already use ntpd.service but in some strange cases it failed, most likely due restoring cluster from snapshot. Now we give more attention to clock setup on cluster, and this problem is mostly solved for us.

Main reason for this bug report is to inform Kubernetes team about huge memory leak issue (936M!) that gradually grows. In flanneld this was due to missing `defer connection.close()`, and i suspect that kubelet has same issue(or something similar)

**Comment by dchen1107** [comment]
@ssergiienko Could you please run cpu profile against Kubelet and post the graph here? Since you are running 1.0.3-X release, we know there are some issues caused by cAdvisor on collecting both disk and network stats on both cpu and memory. 

After 1.0 release, we create daemon resource monitoring related e2e tests, and try to capture those issues as earlier as possible. Also we plan to publish some measurement resource for 1.1 release.

**Comment by ssergiienko** [comment]
Problem appears again. After restore from snapshot i leave cluster for a few hours and when returned i can see that:
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                  
 1922 root      20   0 2578836 1.331g   5480 S 196.0 74.2 218:47.60 kubelet
```
And have this things in logs:
```
Oct 23 15:16:43 some.node.ip kubelet[1922]: E1023 15:16:43.793059    1922 iowatcher.go:108] Unable to decode an event from the watch stream: read tcp some.MASTER.ip:6443: no route to host
Nov 17 18:12:31 some.node.ip kubelet[1922]: E1117 18:12:31.247559    1922 kubelet.go:1933] Error updating node status, will retry: node "some.node.ip" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
Nov 17 18:12:40 some.node.ip kubelet[1922]: E1117 18:12:39.403380    1922 kubelet.go:1933] Error updating node status, will retry: node "some.node.ip" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
Nov 17 18:15:21 some.node.ip kubelet[1922]: E1117 18:15:21.626054    1922 kubelet.go:1933] Error updating node status, will retry: Put https://some.MASTER.ip:6443/api/v1/nodes/some.node.ip/status: net/http: TLS handshake timeout
Nov 17 18:15:37 some.node.ip kubelet[1922]: E1117 18:15:37.672604    1922 kubelet.go:1933] Error updating node status, will retry: error getting node "some.node.ip": Get https://some.MASTER.ip:6443/api/v1/nodes/some.node.ip: net/http: TLS handshake timeout
Nov 17 18:15:44 some.node.ip kubelet[1922]: E1117 18:15:44.154027    1922 kubelet.go:1933] Error updating node status, will retry: node "some.node.ip" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
Nov 17 18:31:25 some.node.ip kubelet[1922]: E1117 18:31:25.700349    1922 reflector.go:183] watch of *api.Service ended with: 401: The event in requested index is outdated and cleared (the requested history has been cleared [777/702]) [1776]
Nov 17 18:33:05 some.node.ip kubelet[1922]: E1117 18:33:05.858930    1922 kubelet.go:1933] Error updating node status, will retry: node "some.node.ip" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
Nov 17 18:33:09 some.node.ip kubelet[1922]: E1117 18:33:09.597084    1922 kubelet.go:1933] Error updating node status, will retry: node "some.node.ip" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
Nov 17 18:34:40 some.node.ip kubelet[1922]: E1117 18:34:40.597862    1922 kubelet.go:1933] Error updating node status, will retry: node "some.node.ip" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
Nov 17 18:34:45 some.node.ip kubelet[1922]: E1117 18:34:45.425611    1922 kubelet.go:1933] Error updating node status, will retry: node "some.node.ip" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
Nov 17 18:36:27 some.node.ip kubelet[1922]: E1117 18:36:26.425660    1922 kubelet.go:1933] Error updating node status, will retry: error getting node "some.node.ip": Get https://some.MASTER.ip:6443/api/v1/nodes/some.node.ip: net/http: TLS handshake timeout
Nov 17 18:36:35 some.node.ip kubelet[1922]: E1117 18:36:35.471640    1922 kubelet.go:1933] Error updating node status, will retry: node "some.node.ip" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
Nov 17 18:38:03 some.node.ip kubelet[1922]: E1117 18:38:03.551200    1922 kubelet.go:1933] Error updating node status, will retry: node "some.node.ip" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
```

But this messages appears only once on this node, and no repeat more, although kubelet continue to eat all resources even now ($date = Tue Nov 17 19:41:37) without new messages in logs.

On our cluster we use ntpd and right now time on nodes and master is correct (at least after some time after restoring from snapshot).

May be useful: etcd version 2.0.9 (on master)

You ask for profiling, but i don't know how to attach running go program and profile it in runtime. But i know strace :) And only one thing the kubelet is doing:
```
strace -v -y -yy -T -s 10000 -p 1922
Process 1922 attached
read(11<anon_inode:inotify>, ^CProcess 1922 detached
 <detached ...>
```
Just this and nothing more. I know, this is not much info, but maybe this helps

**Comment by yujuhong** [comment]
This seems like a cadvisor issue. 

/cc @kubernetes/goog-node

**Comment by cmluciano** [comment]
Is this still an issue that cannot be fixed with [cadvisor housekeeper settings](https://github.com/google/cadvisor/issues/1498#issuecomment-263363359)

**Comment by yujuhong** [comment]
> Is this still an issue that cannot be fixed with cadvisor housekeeper settings

@ssergiienko please file a new issue if that's the case. Thank you.
